doc/specs/#476 - Pinned packages.md

```diff
@@ -0,0 +1,186 @@
+---
+author: Easton Pillay jedieaston/easton@planeteaston.com
+created on: 2022-01-25
+last updated: 2022-02-02
+issue id: 476
+---
+
+# Pinning a package
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality behind "pinning" a package (freezing a package at a certain version so that it will be not be automatically upgraded via `winget upgrade --all`).
+
+## Inspiration
+
+This functionality is inspired by functionality in other package managers, as well as community feedback. Many packages (software development tools and libraries commonly, but other software as well) introduce breaking changes that users may not want integrate into their workflow quite yet.
+
+## Solution Design
+
+A table of packages (by Name, Version and Product Code if available) that are currently pinned will be kept in the local tracking catalog. To be repository independent, these values should be correlated from local sources (Add and Remove Programs, Appx, etc). This also means that the user should be able to pin packages that were not installed via the Windows Package Manager, so that they won't be upgraded via `winget upgrade --all`. 
+
+
+## UI/UX Design
+
+Several commands will need to be added to support pinning packages. Where applicable, equivalent names (or new arguments) for PowerShell cmdlets will also be given.
+
+The main interface for the pinning feature will be a new command, `winget pin`, with a couple of options:
+
+
+```
+winget pin
+PowerShell: Get-WinGetPinnedPackage
+```
+
+Running `winget pin` with no arguments will show a table of currently pinned packages:
+
+```
+Name                                                 Id                                Version    
+------------------------------------------------------------------------------------------------
+Microsoft Bob                                        Microsoft.Bob                     2.35.0
+iTunes                                               Apple.iTunes                      12.0.199.4
+
+```
+
+```
+winget pin --set <package>
+PowerShell: Set-WinGetPinnedPackage
+```
+
+`<package>` correlates to a query, similar to `winget list`. When running this command, the package information will be stored in the table in the tracking catalog. This command should be able to accept Package Identifers (checked against all repositories) as well as names, tags, etc. If the package does not have a version number locally but is available from a source, then the version number from the source should be used.
+
+If there is no package directly available for software a user wants to pin, but it is [included as part of another package](https://github.com/microsoft/winget-cli/issues/1073), the package which includes said piece of software should be treated as pinned unless a newer version doesn't change the version of the pinned software. 
+
+```
+winget pin --clear <package>
```

Comment:
The spec should include discussion of how user driven pinning and `RequiresExplicitUpgrade` interact.  I assert that these are effectively the same thing, with the manifest field driving a default pinned state.  The user should also be able to clear that default pin if they so desire.

Reply:
Maybe this is an opportunity to look at what exactly the RequiresExplicitUpgrade field is supposed to show. If it is intended to show the default pinned state, I would suggest we consider renaming the field to something like `Pinning:` and make it an enum rather than a boolean so that packages could be marked as unable to be pinned without requiring a separate field.

Reply:
> unable to be pinned

That may not be necessary, since winget can't guarantee that a package won't be upgraded anyway, it can only guarantee that `winget upgrade` won't upgrade the package without being explicitly asked.

Reply:
> > unable to be pinned
> 
> That may not be necessary, since winget can't guarantee that a package won't be upgraded anyway, it can only guarantee that `winget upgrade` won't upgrade the package without being explicitly asked.

True, but it would be an opportunity to keep the user informed if we know of specific packages that auto-upgrade. Just an idea, but probably outside the realm of this spec anyways

Reply:
RequiresExplicitUpgrade was about apps that update themselves and was something added to the manifest to somewhat declutter `winget upgrade` for packages that handle upgrades themselves or for ISVs to prefer their own upgrade mechanism. https://github.com/microsoft/winget-cli/issues/1163. The description in the schema probably shouldn't use the word "pinned" as pinning is more of a user driven behavior. Although logically it causes apps to be excluded from `winget upgrade --all` which pinning also does.

Pinning a package is more about a user saying, "I don't want this upgraded". It's often the case for enterprises who have to support specific versions of applications, and developers wanting to have specific versions of packages for their needs.

We may also need to consider the case where packages won't be able to support pinning. Some packages are going to update no matter what "we" want. This could help us to inform users when we can't control the software upgrade mechanism.